### ISSUE TITLE
Identical URLs requested multiple times

### ISSUE THREAD

**Issue by fredcy** [raviqqe/muffet#38]
When I run muffet against a local site I see in the logs that some pages are being requested many times in a single run.  This seem unnecessary and puts extra load on the server being tested.

Here is a simple example. Create "test.html" with this content:
```
<html><body>
<a href="/foo.html">foo</a>
<a href="/test2.html">test2</a>
</body></html>
```
and "test2.html" with this:
```
<html><body>
<a href="/foo.html">foo</a>
</body></html>
```

Then serve this content with `python3 -m http.server`.

And run `muffet http://localhost:8000/test.html`.

The python http.server output I get is this:
```
Serving HTTP on 0.0.0.0 port 8000 (http://0.0.0.0:8000/) ...
127.0.0.1 - - [24/Aug/2018 16:14:01] "GET /test.html HTTP/1.1" 200 -
127.0.0.1 - - [24/Aug/2018 16:14:01] "GET /test2.html HTTP/1.1" 200 -
127.0.0.1 - - [24/Aug/2018 16:14:01] code 404, message File not found
127.0.0.1 - - [24/Aug/2018 16:14:01] "GET /foo.html HTTP/1.1" 404 -
127.0.0.1 - - [24/Aug/2018 16:14:01] code 404, message File not found
127.0.0.1 - - [24/Aug/2018 16:14:01] "GET /foo.html HTTP/1.1" 404 -
```

This shows that "/foo.html" was requested multiple times.

Strangely, small changes to those html files cause different results. If I add a link to test.html, muffet requests foo.html only once in the run.

**Comment by raviqqe** [comment]
Please see #27.

**Comment by fredcy** [comment]
I assume you are referring to the message in https://github.com/raviqqe/muffet/issues/34#issuecomment-395612562 saying that the caching is "best effort". In my usage case that best effort is resulting in hitting the same URLs many many times.  As shown in my test case above, even the most trivial possible case of multiple HTML files having the same link results in multiple hits. Oh well. I'll see what I can do on my fork then.

**Comment by raviqqe** [comment]
Can you try [this branch](https://github.com/raviqqe/muffet/tree/once-per-url) which visits each URL only once?

**Comment by fredcy** [comment]
I just tried that branch and it does seem to solve the test case that I posted above.

When I run it against my real site, however, I'm still seeing some multiple hits to the same URLs. However, even in the master branch that duplicate effort is perhaps not as bad as I thought. I instrumented a fork of the code (as of yesterday's master) to log each cache hit and miss. In a run against my site under test there were 9870 distinct URLs that got a cache miss out of 11192 cache misses over all.  Of the distinct URLs, 408 had more than one cache miss, the highest runners having 34 and 32 misses. I was running with `-c 20` concurrency.  So muffet did 1322 more requests than it needed to get responses for 9870 URLs. That 13% excess is something I can live with.

**Comment by raviqqe** [comment]
Can you tell me the server program and your website you experimented with if you don't mind so that I can test on my machine? The branch should remove any duplicates of requests to the same URLs so I guess the branch has bugs.

**Comment by fredcy** [comment]
Sorry, the web server I'm testing against is not yet public. It will be soon though; cleaning up broken links after a migration is one of our last steps before going live.

**Comment by raviqqe** [comment]
Thank you Fred. I'm glad to hear that.

I run the new branch on my mid-sized website of around 40 pages and it didn't have any problem. But I could see some duplicate logs as you mentioned while they are expected.

For example, although the URLs `https://foo.com/bar`, `https:/foo.com/bar/`, `https://foo.com/bar/index.html`, and `http://foo.com/bar` are all different for `muffet` because they are different URLs basically, web servers might redirect all the URLs other than the first one to the first one depending on its implementation and configuration. As a result, we'll see 4 log lines of accesses to `https://foo.com/bar` while `muffet` of the client side queries for different URLs originally.

So maybe the duplicate accesses you saw were because of those reasons.
